### PR TITLE
Refs and props

### PR DESCRIPTION
**Pull request by koresar**
`state` renamed to `refs`. New aliased `fixed.state` and `state()` were added for `stamp.fixed.refs` and `stamp.refs()`.

Added `stamp.fixed.props`, `stamp.props()` and fourth parameter to the `stampit()` function.

### CODE REVIEW

File: dist/stampit.js [hunk 1]
```diff
@@ -3786,20 +3786,25 @@ function addMethods(fixed, methods) {
...
+function addProps(fixed, propses) {
+  var args = [fixed.props].concat(propses);
+  fixed.props = mixer.merge.apply(null, args);
+  return fixed.props;
+}
 
 function cloneAndExtend(fixed, extensionFunction, args) {
   args = arguments.length > 3 ? slice(arguments, 2, arguments.length) : args;
-  var stamp = stampit(fixed.methods, fixed.state, fixed.enclose);
+  var stamp = stampit(fixed.methods, fixed.refs, fixed.enclose, fixed.props);
```
Comment (ericelliott) [inline dist/stampit.js]: Shouldn't we also support composing with legacy stamps that have `fixed.state`?

Comment (koresar) [inline dist/stampit.js]: Let me add a unit test on that. It's better to be as backwards compatible as possible.

**Comment by koresar** [comment]
I've just addressed the stampit@1.1 compatibility. See the PR please.

File: dist/stampit.js [hunk 2]
```diff
@@ -3829,24 +3835,29 @@ function compose(factories) {
...
+ * @return {Function} factory.props Add deeply cloned properties to the produced objects. Chainable.
+ */
+stampit = function stampit(methods, refs, enclose, props) {
+  var fixed = {methods: {}, refs: {}, enclose: [], props: {}};
+  fixed.state = fixed.refs; // Backward compatibility.
   addMethods(fixed, methods);
-  addState(fixed, state);
+  addRefs(fixed, refs);
   addEnclose(fixed, enclose);
+  addProps(fixed, props);
```
Comment (ericelliott) [inline dist/stampit.js]: This needs to go before the `enclose()` methods so that closures can access stamp props.

File: dist/stampit.js [hunk 3]
```diff
@@ -3909,8 +3933,9 @@ function isStamp(obj) {
   return (
```
Comment (ericelliott) [inline dist/stampit.js]: This will return false for legacy stamps.

File: test/stampit-specs.js [hunk 4]
```diff
@@ -93,70 +93,136 @@ test('stampit().methods(a, b)', function () {
     'Should mixIn objects when multiple methods are passed.');
 });
 
-module('Basics State');
+module('Basics Props');
 
-test('stampit({}, state)', function () {
+test('stampit({}, props)', function () {
```
Comment (ericelliott) [inline test/stampit-specs.js]: Signature is `stampit(methods, refs, enclose, props)`. You have `props` and `refs` swapped in the tests.

File: test/stampit-specs.js [hunk 5]
```diff
@@ -93,70 +93,136 @@ test('stampit().methods(a, b)', function () {
...
   }).create();
 
   ok(obj.a && obj.b,
-    'Should mixIn objects when multiple methods are passed.');
+    'Should mixIn objects when multiple objects are passed.');
+});
+
+module('Basics refs');
+
+test('stampit({}, null, null, refs)', function () {
```
Comment (ericelliott) [inline test/stampit-specs.js]: Signature is `stampit(methods, refs, enclose, props)`. You have `props` and `refs` swapped in the tests.

File: test/stampit-specs.js [hunk 6]
```diff
@@ -443,12 +515,12 @@ test('stampit.convertConstructor() composed', function () {
...
-module('State safety');
+module('Props safety');
 
-test('Stamp state deep cloned for object created', function () {
+test('Stamp props deep cloned for object created', function () {
   var deep = { foo: 'foo', bar: 'bar' };
-  var stamp1 = stampit().state({ deep: deep, foo: 'foo' });
-  var stamp2 = stampit(null, { deep: deep, foo: 'foo' });
+  var stamp1 = stampit().props({ deep: deep, foo: 'foo' });
+  var stamp2 = stampit(null, null, null, { deep: deep, foo: 'foo' });
```
Comment (ericelliott) [inline test/stampit-specs.js]: Signature is `stampit(methods, refs, enclose, props)`. You have `props` and `refs` swapped in the tests.

**Comment by koresar** [comment]
Fixed the mentioned issues. Thanks Eric.
I'm merging this in. It's time. :)

**Pull request merged**